walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +41,223 @@ public int compare(Event a, Event b) {
-      return possibleMeetingTimes;
+    /**
+     * Returns a list of time periods in which the meeting, specified by request, could happen. If
+     * no time exists for all optional and mandatory attendees, find the time slot(s) that allow
+     * mandatory attendees and the greatest possible number of optional attendees to attend.
+     */
+    Collection<TimeRange> resolveBestTime() {
+      ArrayList<TimeRange> mandatoryAttendeesMeetingTimes =
+          getMandatoryAttendeesMeetingTimes(new HashSet<String>(request.getAttendees()));
+      getOptimalMeetingTimes(
+          mandatoryAttendeesMeetingTimes,
+          getChangesInOptionalAttendeesAttendance(
+              getOptionalAttendeesFreeTimes(new HashSet<String>(request.getOptionalAttendees()))));
+
+      // If there are no meeting times with at least one optional attendee, just return
+      // mandatoryAttendeesMeetingTimes.
+      return optimalMeetingTimes.isEmpty() ? mandatoryAttendeesMeetingTimes : optimalMeetingTimes;
     }
-    Collections.sort(events, ORDER_BY_START_ASC);
-
-    // Add first gap.
-    addIfLongEnough(
-        TimeRange.fromStartEnd(0, events.get(0).getWhen().start(), false),
-        possibleMeetingTimes,
-        request.getDuration());
-    int end = events.get(0).getWhen().end();
-    for (Event event : events) {
-      // event can be merged with current time range
-      if (event.getWhen().start() <= end) {
-        end = Math.max(end, event.getWhen().end());
-        continue;
+
+    /**
+     * Returns all meeting times that allow the most optional attendees to attend. These times must
+     * also fall in a time satisfying all mandatory attendees and must be at least minTime long.
+     * Utilizes two pointers: one to mandatoryAttendeesMeetingTimes, one to changeLog.
+     *
+     * @param mandatoryAttendeesMeetingTimes all meeting times satisfying mandatory attendees
+     * @param changes a change log of the number of available optional attendees over time
+     */
+    private void getOptimalMeetingTimes(
+        ArrayList<TimeRange> mandatoryAttendeesMeetingTimes, TreeMap<Integer, Integer> changes) {
+      int mandatoryAttendeesMeetingTimesIndex = 0, prevTime = 0;
+      for (Map.Entry changeEntry : changes.entrySet()) {
+        // First need to back up  mandatoryAttendeesMeetingTimesIndex in case we missed a time range
+        // in mandatoryAttendeesMeetingTimes. Then compare time range from previous time in changeLog
+        // to current time in changeLog with mandatoryAttendeesMeetingTimes that overlap with this
+        // time range.
+        for (mandatoryAttendeesMeetingTimesIndex =
+                Math.max(0, mandatoryAttendeesMeetingTimesIndex - 1);
+            mandatoryAttendeesMeetingTimesIndex < mandatoryAttendeesMeetingTimes.size()
+                && mandatoryAttendeesMeetingTimes.get(mandatoryAttendeesMeetingTimesIndex).start()
+                    < (Integer) changeEntry.getKey();
```

Comment:
Comparing an integer object is different from comparing an `int` variable (and usually not correct). You may need to get the `int` values from both of these and compare those. However, there may be some special behavior of `Integer` that makes this work. Alex can chime in if this is the case.

Reply:
The same goes below (line 121). I realized there may also be auto-conversion from `Integer` to `int` going on here, but either way, it would be more appropriate to get the `int` value of `mandatoryAttendeesMeetingTimes.get(mandatoryAttendeesMeetingTimesIndex).start()` rather than casting the other value to `Integer`.

Reply:
l changed all casts of `changeEntry.getKey()` to `(int)` instead of `(Integer)`. `mandatoryAttendeesMeetingTimes.get(mandatoryAttendeesMeetingTimesIndex).start()` already returns an `int`, but `changeEntry.getKey()` returns an Object.